File: src/weaponinventorymod/gui/WimGuiButtonPoller.java
package weaponinventorymod.gui;

import com.fs.starfarer.api.input.InputEventAPI;

import java.util.ArrayList;
import java.util.List;

final class WimGuiButtonPoller<A> {
    interface ActionHandler<A> {
        void handle(A action);
    }

    private final List<WimGuiButtonBinding<A>> bindings = new ArrayList<WimGuiButtonBinding<A>>();
    private final int framesAfterMouseEvent;
    private int framesRemaining = 0;
    private boolean hasPendingMouseUp = false;
    private boolean suppressNextMouseUp = false;
    private float pendingMouseUpX = 0f;
    private float pendingMouseUpY = 0f;

    WimGuiButtonPoller(int framesAfterMouseEvent) {
        this.framesAfterMouseEvent = Math.max(1, framesAfterMouseEvent);
    }

    List<WimGuiButtonBinding<A>> bindings() {
        return bindings;
    }

    void clearBindings() {
        bindings.clear();
        framesRemaining = 0;
        clearPendingMouseUp();
    }

    void requestPollFromInput(List<InputEventAPI> events) {
        if (events == null) {
            return;
        }
        for (int i = 0; i < events.size(); i++) {
            InputEventAPI event = events.get(i);
            if (event == null) {
                continue;
            }
            if (event.isMouseUpEvent()) {
                if (suppressNextMouseUp) {
                    suppressNextMouseUp = false;
                    continue;
                }
                hasPendingMouseUp = true;
                pendingMouseUpX = event.getX();
                pendingMouseUpY = event.getY();
                framesRemaining = framesAfterMouseEvent;
                return;
            }
            if (event.isMouseDownEvent()) {
                framesRemaining = framesAfterMouseEvent;
                return;
            }
        }
    }

    boolean processIfRequested(ActionHandler<A> handler) {
        if (framesRemaining <= 0 || handler == null) {
            return false;
        }
        framesRemaining--;
        if (hasPendingMouseUp) {
            float x = pendingMouseUpX;
            float y = pendingMouseUpY;
            clearPendingMouseUp();
            for (int i = 0; i < bindings.size(); i++) {
                WimGuiButtonBinding<A> binding = bindings.get(i);
                if (!binding.consumeIfClicked(x, y)) {
                    continue;
                }
                handler.handle(binding.getAction());
                return true;
            }
        }
        for (int i = 0; i < bindings.size(); i++) {
            WimGuiButtonBinding<A> binding = bindings.get(i);
            if (!binding.consumeIfPressed()) {
                continue;
            }
            handler.handle(binding.getAction());
            return true;
        }
        return false;
    }

    void clearCheckedButtons() {
        for (int i = 0; i < bindings.size(); i++) {
            bindings.get(i).clear();
        }
        framesRemaining = 0;
        clearPendingMouseUp();
    }

    void suppressNextMouseUp() {
        suppressNextMouseUp = true;
        framesRemaining = 0;
        clearPendingMouseUp();
    }

    private void clearPendingMouseUp() {
        hasPendingMouseUp = false;
        pendingMouseUpX = 0f;
        pendingMouseUpY = 0f;
    }
}

File: src/main/kotlin/weaponsprocurement/gui/WimGuiButtonPoller.kt
package weaponsprocurement.gui

import com.fs.starfarer.api.input.InputEventAPI
import java.util.ArrayList

class WimGuiButtonPoller<A>(framesAfterMouseEvent: Int) {
    fun interface ActionHandler<A> {
        fun handle(action: A)
    }

    private val bindings: MutableList<WimGuiButtonBinding<A>> = ArrayList()
    private val framesAfterMouseEvent: Int = Math.max(1, framesAfterMouseEvent)
    private var framesRemaining = 0
    private var hasPendingMouseUp = false
    private var suppressNextMouseUp = false
    private var pendingMouseUpX = 0f
    private var pendingMouseUpY = 0f

    fun bindings(): MutableList<WimGuiButtonBinding<A>> = bindings

    fun clearBindings() {
        bindings.clear()
        framesRemaining = 0
        clearPendingMouseUp()
    }

    fun requestPollFromInput(events: List<InputEventAPI>?) {
        if (events == null) {
            return
        }
        for (event in events) {
            if (event.isMouseUpEvent) {
                if (suppressNextMouseUp) {
                    suppressNextMouseUp = false
                    continue
                }
                hasPendingMouseUp = true
                pendingMouseUpX = event.x.toFloat()
                pendingMouseUpY = event.y.toFloat()
                framesRemaining = framesAfterMouseEvent
                return
            }
            if (event.isMouseDownEvent) {
                framesRemaining = framesAfterMouseEvent
                return
            }
        }
    }

    fun processIfRequested(handler: ActionHandler<A>?): Boolean {
        if (framesRemaining <= 0 || handler == null) {
            return false
        }
        framesRemaining--
        if (hasPendingMouseUp) {
            val x = pendingMouseUpX
            val y = pendingMouseUpY
            clearPendingMouseUp()
            for (binding in bindings) {
                if (!binding.consumeIfClicked(x, y)) {
                    continue
                }
                handler.handle(binding.getAction())
                return true
            }
        }
        for (binding in bindings) {
            if (!binding.consumeIfPressed()) {
                continue
            }
            handler.handle(binding.getAction())
            return true
        }
        return false
    }

    fun clearCheckedButtons() {
        for (binding in bindings) {
            binding.clear()
        }
        framesRemaining = 0
        clearPendingMouseUp()
    }

    fun suppressNextMouseUp() {
        suppressNextMouseUp = true
        framesRemaining = 0
        clearPendingMouseUp()
    }

    private fun clearPendingMouseUp() {
        hasPendingMouseUp = false
        pendingMouseUpX = 0f
        pendingMouseUpY = 0f
    }
}
